### PR TITLE
Replace ArgumentNullException(nameof()) -> ArgumentNullException.ThrowIfNull() 2

### PR DESCRIPTION
**Pull request by CarloToso**
# PR Summary
Continues #18784

8 search results for `throw new ArgumentNullException(`, I don't think they should be changed

## PR Context

The new API [ArgumentNullException.ThrowIfNull](https://learn.microsoft.com/en-us/dotnet/api/system.argumentnullexception.throwifnull?view=net-7.0#system-argumentnullexception-throwifnull(system-object-system-string)) allows you to omit the parameter name, in which case it will use the name of the argument passed in. The passed in argument usually is the parameter to be validated, so that's just the right name to use when calling `ArgumentNullException.ThrowIfNull(argument)`.

### CODE REVIEW

File: src/Microsoft.Management.UI.Internal/ManagementList/Common/WpfHelp.cs [hunk 1]
```diff
@@ -214,10 +214,7 @@ public static void AddChild(FrameworkElement parent, FrameworkElement element)
         {
             ArgumentNullException.ThrowIfNull(element);
 
-            if (parent == null)
-            {
-                throw new ArgumentNullException("element");
-            }
+            ArgumentNullException.ThrowIfNull(parent, "element");
```
Comment (iSazonov) [inline src/Microsoft.Management.UI.Internal/ManagementList/Common/WpfHelp.cs:217]:
```suggestion
            ArgumentNullException.ThrowIfNull(parent, nameof(element));
```

File: src/Microsoft.Management.UI.Internal/ManagementList/Common/WpfHelp.cs [hunk 2]
```diff
@@ -370,10 +367,7 @@ public static T FindVisualAncestorData<T>(this DependencyObject obj)
         /// <exception cref="ArgumentNullException">The specified value is a null reference.</exception>
         public static T FindVisualAncestor<T>(this DependencyObject @object) where T : class
         {
-            if (@object == null)
-            {
-                throw new ArgumentNullException("object");
-            }
+            ArgumentNullException.ThrowIfNull(@object, "object");
```
Comment (iSazonov) [inline src/Microsoft.Management.UI.Internal/ManagementList/Common/WpfHelp.cs:370]:
```suggestion
            ArgumentNullException.ThrowIfNull(@object, nameof(@object));
```

File: src/Microsoft.Management.UI.Internal/ManagementList/ManagementList/DefaultStringConverter.cs [hunk 3]
```diff
@@ -62,7 +62,9 @@ public string DefaultValue
         /// </remarks>
         public object Convert(object[] values, Type targetType, object parameter, System.Globalization.CultureInfo culture)
         {
-            if (values == null || values.Length != 1)
+            ArgumentNullException.ThrowIfNull(values);
+
+            if (values.Length != 1)
             {
                 throw new ArgumentNullException("values");
```
Comment (iSazonov) [inline src/Microsoft.Management.UI.Internal/ManagementList/ManagementList/DefaultStringConverter.cs:69]: I guess it is a candidate for ArgumentRangeException.

File: src/Microsoft.Management.UI.Internal/ShowCommand/ViewModel/AllModulesViewModel.cs [hunk 4]
```diff
@@ -79,11 +79,13 @@ public class AllModulesViewModel : INotifyPropertyChanged
         /// <param name="commands">Commands to show.</param>
         public AllModulesViewModel(Dictionary<string, ShowCommandModuleInfo> importedModules, IEnumerable<ShowCommandCommandInfo> commands)
         {
-            if (commands == null || !commands.GetEnumerator().MoveNext())
+            ArgumentNullException.ThrowIfNull(commands);
+
+            if (!commands.GetEnumerator().MoveNext())
             {
                 throw new ArgumentNullException("commands");
```
Comment (iSazonov) [inline src/Microsoft.Management.UI.Internal/ShowCommand/ViewModel/AllModulesViewModel.cs:86]: The same.

File: src/Microsoft.Management.UI.Internal/ShowCommand/ViewModel/AllModulesViewModel.cs [hunk 5]
```diff
@@ -79,11 +79,13 @@ public class AllModulesViewModel : INotifyPropertyChanged
...
             }
-
+
```
Comment (iSazonov) [inline src/Microsoft.Management.UI.Internal/ShowCommand/ViewModel/AllModulesViewModel.cs:88]: Please remove extra spaces.

File: src/Microsoft.PowerShell.Commands.Management/cimSupport/cmdletization/SessionBasedWrapper.cs [hunk 6]
```diff
@@ -82,7 +82,7 @@ protected TSession[] Session
 
             set
             {
-                _session = value ?? throw new ArgumentNullException(nameof(value));
+                _session = value ?? ArgumentNullException.ThrowIfNull(value);
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Management/cimSupport/cmdletization/SessionBasedWrapper.cs:85]: Is it correct?

File: src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs [hunk 7]
```diff
@@ -578,7 +578,7 @@ private void ProcessMTUSize(string targetNameOrAddress)
                 WriteObject(new PingMtuStatus(
                     Source,
                     resolvedTargetName,
-                    replyResult ?? throw new ArgumentNullException(nameof(replyResult)),
+                    replyResult ?? ArgumentNullException.ThrowIfNull(replyResult),
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs:581]: The same.

**Comment by iSazonov** [comment]
We get ArgumentOutOfRangeException.ThrowIfZero() only with .Net 8.0 in next January.

**Comment by CarloToso** [comment]
What can I do about this build error? `'ArgumentOutOfRangeException' does not contain a definition for 'ThrowIfZero'`

File: src/Microsoft.PowerShell.Commands.Management/cimSupport/cmdletization/SessionBasedWrapper.cs [hunk 8]
```diff
@@ -82,7 +82,7 @@ protected TSession[] Session
 
             set
             {
-                _session = value ?? ArgumentNullException.ThrowIfNull(value);
+                _session = value ?? throw new ArgumentNullException(nameof(value));
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Management/cimSupport/cmdletization/SessionBasedWrapper.cs:85]:
```suggestion
                ArgumentNullException.ThrowIfNull(value);
                _session = value;
```

File: src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs [hunk 9]
```diff
@@ -578,7 +578,7 @@ private void ProcessMTUSize(string targetNameOrAddress)
                 WriteObject(new PingMtuStatus(
                     Source,
                     resolvedTargetName,
-                    replyResult ?? ArgumentNullException.ThrowIfNull(replyResult),
+                    replyResult ?? throw new ArgumentNullException(nameof(replyResult)),
                     CurrentMTUSize));
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs:582]: The same.

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs [hunk 10]
```diff
@@ -8025,15 +8025,9 @@ internal static bool CreateJunction(string path, string target)
         [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Reliability", "CA2001:AvoidCallingProblematicMethods")]
         private static bool WinCreateJunction(string path, string target)
         {
-            if (string.IsNullOrEmpty(path))
-            {
-                throw new ArgumentNullException(nameof(path));
-            }
+            ArgumentException.ThrowIfNullOrEmpty(path);
```
Comment (iSazonov) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs:8028]: Please fix test_FileSystemProvider.cs:line 87

Comment (CarloToso) [reply src/System.Management.Automation/namespaces/FileSystemProvider.cs:8028]: Is this the right fix? `Assert.Throws<System.ArgumentNullException>`  -> `Assert.Throws<System.ArgumentException>`

Comment (iSazonov) [reply src/System.Management.Automation/namespaces/FileSystemProvider.cs:8028]: Yes, it is right.

File: src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs (hunk 9 above)
Comment (CarloToso) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs]: Done, any insights on the failing test?

**Comment by iSazonov** [comment]
@CarloToso Please remove all with IsNullOrEmpty. I see we use ArgumentNullException but then will raise two exceptions (for null and for empty) instead of one. This can be a breaking change (specially in remoting code like WSMan). It is better to discuss with MSFT team before.
```c#
            if (string.IsNullOrEmpty(paramName))
            {
                throw new ArgumentNullException(nameof(paramName));
            }
```

File: src/System.Management.Automation/engine/Utils.cs [hunk 11]
```diff
@@ -1740,10 +1740,7 @@ internal static class Requires
     {
         internal static void NotNull(object value, string paramName)
         {
-            if (value == null)
-            {
-                throw new ArgumentNullException(paramName);
-            }
+            ArgumentNullException.ThrowIfNull(value, paramName);
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/Utils.cs:1743]: In follow PR we could replace all `NotNull` with the `ArgumentNullException.ThrowIfNull`.

File: src/Microsoft.WSMan.Management/ConfigProvider.cs [hunk 12]
```diff
@@ -3413,7 +3413,13 @@ private static string NormalizePath(string path, string host)
...
         private PSObject GetItemValue(string path)
         {
-            if (string.IsNullOrEmpty(path) || (path.Length == 0))
+            if (string.IsNullOrEmpty(path))
+            {
+                throw new ArgumentNullException(path);
+            }
+
+            // in .NET8.0 -> ArgumentOutOfRangeException.ThrowIfZero(path.Length);
+            if (path.Length == 0)
```
Comment (iSazonov) [inline src/Microsoft.WSMan.Management/ConfigProvider.cs:3422]:
```suggestion
            if (string.IsNullOrEmpty(path))
```

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs [hunk 13]
```diff
@@ -7874,16 +7874,14 @@ internal static bool CreateJunction(string path, string target)
         {
 #if UNIX
             return false;
-#else
-            if (string.IsNullOrEmpty(path))
-            {
-                throw new ArgumentNullException(nameof(path));
-            }
+        }
```
Comment (iSazonov) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs:7877]: Please revert this and below.

File: test/xUnit/csharp/test_FileSystemProvider.cs [hunk 14]
```diff
@@ -84,7 +84,7 @@ public void TestCreateJunctionFails()
             }
             else
             {
-                Assert.Throws<System.ArgumentNullException>(delegate { InternalSymbolicLinkLinkCodeMethods.CreateJunction(string.Empty, string.Empty); });
+                Assert.Throws<System.ArgumentException>(delegate { InternalSymbolicLinkLinkCodeMethods.CreateJunction(string.Empty, string.Empty); });
```
Comment (iSazonov) [inline test/xUnit/csharp/test_FileSystemProvider.cs:87]: Please revert.

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs [hunk 15]
```diff
@@ -7884,7 +7884,7 @@ internal static bool CreateJunction(string path, string target)
             {
                 throw new ArgumentNullException(nameof(target));
             }
-
+
```
Comment (iSazonov) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs:7887]:
```suggestion

```

File: src/System.Management.Automation/utils/ParserException.cs [hunk 16]
```diff
@@ -131,7 +131,10 @@ internal ParseException(string message, string errorId, Exception innerException
             Justification = "ErrorRecord is not overridden in classes deriving from ParseException")]
         public ParseException(Language.ParseError[] errors)
         {
-            if ((errors == null) || (errors.Length == 0))
+            ArgumentNullException.ThrowIfNull(errors);
+
+            // in .NET8.0 -> ArgumentOutOfRangeException.ThrowIfZero(errors.Length);
+            if (errors.Length == 0)
```
Comment (iSazonov) [inline src/System.Management.Automation/utils/ParserException.cs:137]:
```suggestion
            ArgumentNullException.ThrowIfNull(errors);

            if (errors.Length == 0)
```

File: src/System.Management.Automation/utils/BackgroundDispatcher.cs [hunk 17]
```diff
@@ -70,7 +70,7 @@ public BackgroundDispatcher(EventProvider transferProvider, EventDescriptor tran
         // internal for unit testing only.  Otherwise, would be private.
         internal BackgroundDispatcher(IMethodInvoker etwActivityMethodInvoker)
         {
-            _etwActivityMethodInvoker = etwActivityMethodInvoker ?? throw new ArgumentNullException(nameof(etwActivityMethodInvoker));
+            _etwActivityMethodInvoker = etwActivityMethodInvoker ?? ArgumentNullException.ThrowIfNull(etwActivityMethodInvoker);
```
Comment (iSazonov) [inline src/System.Management.Automation/utils/BackgroundDispatcher.cs:73]:
```suggestion
            ArgumentNullException.ThrowIfNull(etwActivityMethodInvoker);
            _etwActivityMethodInvoker = etwActivityMethodInvoker;
```

File: src/Microsoft.Management.UI.Internal/ManagementList/ManagementList/DefaultStringConverter.cs [hunk 18]
```diff
@@ -62,9 +62,11 @@ public string DefaultValue
         /// </remarks>
         public object Convert(object[] values, Type targetType, object parameter, System.Globalization.CultureInfo culture)
         {
-            if (values == null || values.Length != 1)
+            ArgumentNullException.ThrowIfNull(values);
+
+            if (values.Length != 1)
             {
-                throw new ArgumentNullException("values");
+                throw new ArgumentOutOfRangeException("values");
```
Comment (iSazonov) [inline src/Microsoft.Management.UI.Internal/ManagementList/ManagementList/DefaultStringConverter.cs:69]:
```suggestion
                throw new ArgumentNullException("values");
```

File: src/Microsoft.Management.UI.Internal/ShowCommand/ViewModel/AllModulesViewModel.cs [hunk 19]
```diff
@@ -79,9 +79,11 @@ public class AllModulesViewModel : INotifyPropertyChanged
         /// <param name="commands">Commands to show.</param>
         public AllModulesViewModel(Dictionary<string, ShowCommandModuleInfo> importedModules, IEnumerable<ShowCommandCommandInfo> commands)
         {
-            if (commands == null || !commands.GetEnumerator().MoveNext())
+            ArgumentNullException.ThrowIfNull(commands);
+
+            if (!commands.GetEnumerator().MoveNext())
             {
-                throw new ArgumentNullException("commands");
+                throw new ArgumentOutOfRangeException("commands");
```
Comment (iSazonov) [inline src/Microsoft.Management.UI.Internal/ShowCommand/ViewModel/AllModulesViewModel.cs:86]:
```suggestion
                throw new ArgumentNullException("commands");
```

File: src/System.Management.Automation/utils/perfCounters/CounterSetRegistrarBase.cs [hunk 20]
```diff
@@ -107,8 +107,11 @@ public abstract class CounterSetRegistrarBase
             CounterSetId = counterSetId;
             CounterSetInstType = counterSetInstType;
             CounterSetName = counterSetName;
-            if ((counterInfoArray == null)
-                || (counterInfoArray.Length == 0))
+
+            ArgumentNullException.ThrowIfNull(counterInfoArray);
+
+            // in .NET8.0 -> ArgumentOutOfRangeException.ThrowIfZero(counterInfoArray.Length);
+            if (counterInfoArray.Length == 0)
```
Comment (iSazonov) [inline src/System.Management.Automation/utils/perfCounters/CounterSetRegistrarBase.cs:114]:
```suggestion

            ArgumentNullException.ThrowIfNull(counterInfoArray);

            if (counterInfoArray.Length == 0)
```

**Review by iSazonov — APPROVED** [review]

File: src/Microsoft.Management.UI.Internal/ManagementList/Common/WpfHelp.cs [hunk 21]
```diff
@@ -214,10 +214,7 @@ public static void AddChild(FrameworkElement parent, FrameworkElement element)
         {
             ArgumentNullException.ThrowIfNull(element);
 
-            if (parent == null)
-            {
-                throw new ArgumentNullException("element");
-            }
+            ArgumentNullException.ThrowIfNull(parent, nameof(element));
```
Comment (xtqqczze) [inline src/Microsoft.Management.UI.Internal/ManagementList/Common/WpfHelp.cs:217]: @xtqqczze The second parameter appears to be incorrect.

Comment (CarloToso) [reply src/Microsoft.Management.UI.Internal/ManagementList/Common/WpfHelp.cs:217]: @iSazonov suggested this change https://github.com/PowerShell/PowerShell/pull/18792#discussion_r1048671618

Comment (iSazonov) [reply src/Microsoft.Management.UI.Internal/ManagementList/Common/WpfHelp.cs:217]: The code is frozen and we should keep its behavior unchanged.

**Pull request merged**